provider/macros/src/lib.rs

```diff
@@ -51,43 +58,71 @@ fn data_struct_impl(attr: AttributeArgs, item: ItemStruct) -> TokenStream2 {
         .to_compile_error();
     }
 
-    let docs = format!("Marker type for [`{}`]", name);
+    let mut attr_it = attr.into_iter().peekable();
 
-    let mut result = if lifetimes.get(0).is_some() {
+    let mut result = if let Some(NestedMeta::Lit(key_lit)) =
+        attr_it.next_if(|v| matches!(v, NestedMeta::Lit(_)))
+    {
+        let key_str = match &key_lit {
+            syn::Lit::Str(lit_str) => lit_str.value(),
+            _ => panic!("Key must be a string"),
+        };
+        let docs = format!("Marker type for [`{}`]: \"{}\"", name, key_str);
         quote!(
             #[doc = #docs]
             pub struct #marker;
-
             impl icu_provider::DataMarker for #marker {
-                type Yokeable = #name<'static>;
+                type Yokeable = #name_with_lt;
+            }
+            impl icu_provider::ResourceMarker for #marker {
+                const KEY: icu_provider::ResourceKey = icu_provider::resource_key!(#key_lit);
             }
-
-            #[derive(Yokeable, ZeroCopyFrom)]
-            #item
         )
     } else {
+        let docs = format!("Marker type for [`{}`]", name);
         quote!(
             #[doc = #docs]
             pub struct #marker;
-
             impl icu_provider::DataMarker for #marker {
-                type Yokeable = #name;
+                type Yokeable = #name_with_lt;
             }
-
-            #[derive(Yokeable, ZeroCopyFrom)]
-            #item
         )
     };
 
-    if let Some(attr0) = attr.get(0) {
-        if let NestedMeta::Lit(key_str) = attr0 {
-            result.extend(quote!(
-                impl icu_provider::ResourceMarker for #marker {
-                    const KEY: icu_provider::ResourceKey = icu_provider::resource_key!(#key_str);
-                }
-            ));
+    for single_attr in attr_it {
+        match single_attr {
+            NestedMeta::Meta(Meta::NameValue(name_value)) => {
+                let extra_marker = &name_value.path;
+                let key_lit = &name_value.lit;
+                let key_str = match key_lit {
+                    syn::Lit::Str(lit_str) => lit_str.value(),
+                    _ => panic!("Key must be a string"),
+                };
+                let docs = format!("Marker type for key \"{}\": [`{}`]", key_str, name);
```

Comment:
Are we keeping keys as client-facing concepts though? If they correspond to `ResourceMarker`s, will a client ever encounter a key?

Reply:
I guess the "key" term is used widely in datagen and data pipeline documentation. I'll rephrase my question then: do we need both `ResourceKey` and `ResourceMarker` to be distinct client-facing concepts?

Reply:
@robertbastian  Yes, because `ResourceKey` is still used in the other provider traits (BufferProvider, AnyProvider, and DynProvider).  It is elided only in ResourceProvider.

Reply:
Aren't those usages all internal?

Reply:
The four traits are all public and can be implemented by users.

Reply:
I'm marking this conversation as resolved; if it is not actually resolved, please reopen it